EIPS/eip-memory-paging.md

```diff
@@ -0,0 +1,276 @@
+---
+title: Linear, Page-Based Memory Costing
+description: Linearize Memory Costing and replace the current quadratic formula with a page-based cost model.
+author: Charles Cooper (@charles-cooper)
+discussions-to: https://ethereum-magicians.org/t/eip-linearize-memory-costing/23290
+status: Draft
+type: Standards Track
+category: Core
+created: 2025-03-27
+requires:
+---
+
+## Abstract
+
+This EIP replaces the quadratic memory model in the EVM with a linear, page-based costing model. Memory is virtually addressable. Page allocation and page thrashing are included in the cost model. After applying this EIP, memory limits are invariant to the state of the message call stack.
+
+## Motivation
+
+The EVM currently uses a quadratic pricing model for its memory. This was originally put in place to defend against DoS attacks. However, the memory model has several drawbacks.
+
+1. It is anachronistic. Even at a gas limit of 30 million gas, users can only use 3MB of memory (and that burns all gas). Since the quadratic term for memory expansion starts kicking in at 724 bytes, users use much less memory in practice. To use even 64KB of memory -- the amount of memory available to a PC from the early 80's -- costs 14336 gas!
+2. The quadratic model makes it difficult to reason about how much memory a transaction can allocate. It requires solving an optimization problem which involves computing how many message calls are available to recurse into based on the call stack limit (and, post [EIP-150](./eip-150.md), the 63/64ths rule), and then maximizing the memory used per message call.
+3. The quadratic model makes it impossible for high-level smart contracts languages to get the benefits of virtual memory. Most modern programming languages maintain what is known as the "heap" and the "call stack". The heap is used to allocate objects which live past the lifetime of their current function frame, whereas the call stack is used to allocate objects which live in the current function frames. Importantly, the call stack starts at the top of memory and grows down, while the heap starts at the bottom of memory and grows up, thus the language implementation does not need to worry about the two regions of memory interfering with each other. This is a feature which is enabled by virtual, paged memory, which has been present in operating systems since the early 90's. However, smart contract languages like Vyper and Solidity are not able to implement this, leading to inefficiencies in their memory models.
+
+This EIP proposes a linear costing model which more closely reflects the hardware of today. It uses a virtual addressing schema so that memory pages are not allocated until they are actually accessed. Notably, the data structures used for costing memory do not need to be part of the memory implementation itself, which suggests an elegant implementation using `mmap`.
+
+## Specification
```

Comment:
Question: 

Is this an accurate summary of the page expansion cost: 

1. New pages cost 100 gas to expand. 
2. MLOAD/MSTORE cost 3 gas for "hot" pages (within last 512 accessed pages).
3. MLOAD/MSTORE cost 6 gas for "cold" pages.
4. Max "Hot" memory size is 2097152.

Reply:
1. yes
2. yes
3. i wrote it as 6 + 3 == 9 gas, but according to the benchmarks, 6 gas technically also would pay for the thrash
4. yes